Update surepy to v0.4.0

insertjokehere opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Update surepy library used by the Sure Petcare integration to v0.4.0 - this is required to support setting flap lock states in a future MR

## Additional information
* SurePy change diff - https://github.com/benleb/surepy/compare/v0.2.6...v0.4.0
* Changelog in GitHub - https://github.com/benleb/surepy/releases

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

insertjokehere committed:
Update surepy to v0.4.0

MartinHjelmare commented:
Please add a link to a changelog or GitHub commit compare view for the version bump in the PR description. Thanks!

insertjokehere commented:
@MartinHjelmare description updated. Any suggestions about the CI failure?

MartinHjelmare commented:
The surepy library now requires Python 3.8+. We still require Python 3.7 compatibility. Although we're planning to soon move to minimum Python 3.8.

MartinHjelmare commented:
I suggest we close this for now until Home Assistant drops Python 3.7.

insertjokehere commented:
@MartinHjelmare ah, that would explain it. Do you know what the timeline for actually dropping support for 3.7 is? The [0.116 release notes](https://www.home-assistant.io/blog/#python-37-deprecated) said it would be dropped in 0.118, but that doesn't seem to actually have happened?

MartinHjelmare commented:
We're working on it.

https://github.com/home-assistant/core/pull/43805

There's no timeline. Just asap.

insertjokehere commented:
Awesome, thanks :smile:

frenck commented:
The 3.7 support has been dropped on our dev branch. If you rebase this PR, the builds should pass.

MartinHjelmare commented:
Python 3.8 has been dropped. Please rebase this PR in latest dev branch to let the build pass.

frenck commented:
I was rebasing some PRs anyways, so took the liberty of rebasing this one as well.

MartinHjelmare commented on homeassistant/components/surepetcare/__init__.py:
```diff
@@ -158,8 +159,9 @@ def __init__(self, hass, surepy: SurePetcare, ids: List[Dict[str, Any]]) -> None
-
+        # Fetch all data from SurePet API, refreshing the surepy cache
+        # TODO: get surepy upstream to add a method to clear the cache explicitly
+        await self.surepy._get_resource(resource=MESTART_RESOURCE)
```
This will probably fail our linter. It's ok to disable it locally here for now.

connectionjewelry commented:
Interesting please show me and pr my store [http://trendsswaroski.myshopify.com](www.connectionjewelry.com![Uploading 103443500...])

frenck commented:
Please don't squash changes, it makes it really hard to follow progress for a reviewer. Instead, just commit and push each change. We'll squash and merge this whole PR in the end. Thanks 👍

insertjokehere commented:
Sorry!

Changes since last review:
* Configured pylint to ignore a TODO, and the private member access that the TODO was talking about
* Fixed test failure due to a bad mock

MartinHjelmare commented on homeassistant/components/surepetcare/__init__.py:
```diff
@@ -158,8 +159,11 @@ def __init__(self, hass, surepy: SurePetcare, ids: List[Dict[str, Any]]) -> None
-        await self.surepy.get_data()
-
+        # Fetch all data from SurePet API, refreshing the surepy cache
+        # TODO: get surepy upstream to add a method to clear the cache explicitly pylint: disable=W0511
+        await self.surepy._get_resource(  # pylint: disable=W0212
```
```suggestion
        # TODO: get surepy upstream to add a method to clear the cache explicitly pylint: disable=fixme
        await self.surepy._get_resource(  # pylint: disable=protected-access
```

MartinHjelmare committed:
Clarify pylint disable

MartinHjelmare reviewed (approved):
Thanks!
